### ISSUE TITLE
[2.1|3.0] databaseSession_lifetime is ignored if session.gc_maxlifetime is higher

### ISSUE THREAD

**Issue by arlomedia** [SimpleMachines/SMF#8134]
### Basic Information

I moved my forum to a new server that has session.gc_maxlifetime set to one week in php.ini. My SMF sessions table then grew so large that its cleanup queries were taking 10+ minutes and blocking other connections. I see that the Cookies and Sessions settings page has a "Seconds before an unused session timeout" setting (mine is set to 2880), which corresponds to a databaseSession_lifetime variable. But in the sessionGC() function in Session.php, that gets ignored if it is lower than session.gc_maxlifetime.

I see why a lower value would be ignored in that function, because that function will only be called as often as session.gc_maxlifetime anyway. But it isn't mentioned on the settings page or in the help text that this setting will be ignored. I think you could resolve this by setting session.gc_maxlifetime to the databaseSession_lifetime value in the loadSession() function. Then SMF would use the "Seconds before an unused session timeout" setting regardless of the php.ini settings.

BTW, the one week session.gc_maxlifetime value in php.ini is appropriate for the other applications on my server, so the ability to override it specifically for SMF would be ideal. Since that setting is already displayed in SMF, we're almost there, and just need to use that value in all cases.

### Steps to reproduce

1. Set "Seconds before an unused session timeout" in SMF to 2880.
2. Set session.gc_maxlifetime to 604800 in php.ini.
3. Run this SQL query in the database: SELECT COUNT(*) FROM sessions WHERE last_update<UNIX_TIMESTAMP()-2880;

### Expected result

The number of sessions more than 2880 seconds old would remain fairly small, resetting each time the garbage collection runs.

### Actual result

The number of sessions more than 2880 seconds old grows continuously for a week at a time.

### Version/Git revision

2.1.4

### Database Engine

MySQL

### Database Version

8.0.30

### PHP Version

8.0.30

**Comment by jdarwood007** [comment]
What you're asking is already done in the code:
https://github.com/SimpleMachines/SMF/blob/0c3bd52261990efa9a91ffe9b7c5ccebf24be9c8/Sources/Session.php#L77-L78

Unless I have read your issue wrong.

**Comment by arlomedia** [comment]
The code you quoted doesn't address the issue. That makes SMF use its databaseSession_lifetime if session.gc_maxlifetime is too small, which is fine. But I want SMF to use its databaseSession_lifetime  whenever it is set, including when it's smaller than session.gc_maxlifetime. I suggested a way to do that in my second paragraph above.

**Comment by jdarwood007** [comment]
https://github.com/SimpleMachines/SMF/blob/0c3bd52261990efa9a91ffe9b7c5ccebf24be9c8/Sources/Session.php#L231-L241

What you're suggesting is that we do:
```
			$max_lifetime = max(min($modSettings['databaseSession_lifetime'], (int) ini_get('session.gc_maxlifetime')), 60);
```

This is to first get the lower of the smf or server session lifetime and then ensure it is greater than 60 seconds.

**Comment by arlomedia** [comment]
That also wouldn't fix the problem because that code only runs when databaseSession_lifetime  is more than session.gc_maxlifetime. I need to use a value smaller than the session.gc_maxlifetime of 604800 in php.ini. The current code ignores my databaseSession_lifetime (set in the admin interface as "Seconds before an unused session timeout") and defers to the default session.gc_maxlifetime in that situation.

Here's how I fixed it in my installation. In loadSession(), after:

`
@ini_set('session.gc_probability', '1');
`

I added:

`
@ini_set('session.gc_maxlifetime', max($modSettings['databaseSession_lifetime'], 60));
`

Now it doesn't matter if the code uses session.gc_maxlifetime instead of databaseSession_lifetime because they're the same.

Note that I have databaseSession_enable (set in the admin interface as "Use database driven sessions") turned on. Users who have that turned off would still have the problem from two places:

In loadSession(), databaseSession_lifetime is ignored when databaseSession_enable is off and session.gc_maxlifetime is more than 1440. I don't know why this setting wouldn't be used all the time, regardless of whether databaseSession_enable is on or off, and regardless of the default session.gc_maxlifetime setting. To do that, you could remove this conditional block and place the new line I added after the `if (!empty($modSettings['databaseSession_enable']))` block, so all users get the fix regardless of their settings.

In gc(), databaseSession_lifetime is only used when it's more than session.gc_maxlifetime. If you made the change in the previous paragraph, I think you could remove that conditional block and just use $max_lifetime directly.

**Comment by arlomedia** [comment]
Here's a pull request:

https://github.com/SimpleMachines/SMF/pull/8704

I might have done that wrong; I didn't see a way to create it from this issue thread.

**Comment by jdarwood007** [comment]
Nope, that is how you do pull requests in GitHub.  You did submit it against 3.0, so it won't be in 2.1 unless we backport it or another PR is submitted for 2.1.

What I see the PR doing is always setting the session lifetime in the GC, even if we didn't load an existing session.  So what you are saying is that GC has been called and its triggering with the PHP defaults instead of what we specified in the database session max lifetime?

The code that was completely removed to provide the fallback $max_lifetime if it wasn't, seems to me to be a sanity check.  I went through our easy code history in git, but I can't find why it was added (and older history is in svn, so harder to search).  But I think it should remain, if not updated to handle the cases better, if not simply to ensure we are working with something sane.

**Comment by arlomedia** [comment]
The relevant code is the same in 2.1 and 3.0, so you could do the same change in both if you're still maintaining 2.1.

Yes, until I made this patch in my installation, GC was called with PHP's very high default max_lifetime, which allowed the session table to grow so large that it killed my system performance.

Checking for missing or invalid settings is fine. I think the important thing is to remove the conditionals described above that cause the databaseSession_lifetime setting to be ignored for no apparent reason. It is a useful setting if the code actually uses it. :-)
